synapse/lib/base.py

```diff
@@ -0,0 +1,571 @@
+
+        '''
+        event = (evtname, info)
+        if self.isfini:
+            return event
+
+        await self.dist(event)
+        return event
+
+    async def dist(self, mesg):
+        '''
+        Distribute an existing event tuple.
+
+        Args:
+            mesg ((str,dict)):  An event tuple.
+
+        Example:
+
+            base.dist( ('foo',{'bar':'baz'}) )
+
+        '''
+        if self.isfini:
+            return ()
+
+        ret = []
+        for func, filt in self._syn_funcs.get(mesg[0], ()):
+
+            try:
+
+                if any(True for k, v in filt if mesg[1].get(k) != v):
+                    continue
+
+                retn = func(mesg)
+                if asyncio.iscoroutine(retn):
+                    retn = await retn
+                ret.append(retn)
+
+            except Exception as e:
+                logger.exception('base %s error with mesg %s', self, mesg)
+
+        for func in self._syn_links:
+            try:
+                ret.append(await func(mesg))
+            except Exception as e:
+                logger.exception('base %s error with mesg %s', self, mesg)
+
+        return ret
+
+    async def fini(self):
+        '''
+        Shut down the object and notify any onfini() coroutines.
+
+        Returns:
+            Remaining ref count
+        '''
+        if self.isfini:
+            return
+
+        self._syn_refs -= 1
+        if self._syn_refs > 0:
+            return self._syn_refs
+
+        self.isfini = True
+
+        fevt = self.finievt
+        for fini in self._fini_funcs:
+
+            try:
+                rv = fini()
+                if asyncio.iscoroutine(rv):
+                    await rv
+
+            except Exception as e:
+                logger.exception('fini failed')
+
+        if fevt is not None:
+            fevt.set()
+        self._syn_funcs.clear()
+        del self._fini_funcs[:]
+        return 0
+
+    async def waitfini(self, timeout=None):
+        '''
+        Wait for the event bus to fini()
+
+        Example:
+
+            bus.waitfini(timeout=30)
+
+        '''
+
+        if self.isfini:
+            return True
+
+        if self.finievt is None:
+            self.finievt = asyncio.Event(loop=self.loop)
+
+        try:
+            await asyncio.wait_for(self.finievt.wait(), timeout, loop=self.loop)
+        except asyncio.TimeoutError:
+            return None
+        return True
+
+    def main(self):
+        '''
+        Helper function to block until shutdown ( and handle ctrl-c and SIGTERM).
+
+        Examples:
+            Run a event bus, wait until main() has returned, then do other stuff::
+
+                foo = EventBus()
+                foo.main()
+                dostuff()
+
+        Notes:
+            This does fire a 'ebus:main' event prior to entering the
+            waitfini() loop.
+
+        Returns:
+            None
+        '''
+        assert self.loop == s_glob.plex.loop
```

Comment:
assert can be compiled away. what is the intent here?

Reply:
just a sanity check. Actually, I can remove it, since we're no longer strongly tied to the plex loop.